core/src/ast/data/nested_meta.rs

```diff
@@ -0,0 +1,135 @@
+                segments.push_punct(punct);
+            }
+            if segments.is_empty() {
+                return Err(input.parse::<Ident>().unwrap_err());
+            } else if segments.trailing_punct() {
+                return Err(input.error("expected path segment after `::`"));
+            }
+            segments
+        },
+    })
+}
+
+fn parse_meta_after_path<'a>(path: Path, input: ParseStream<'a>) -> syn::Result<Meta> {
+    if input.peek(token::Paren) || input.peek(token::Bracket) || input.peek(token::Brace) {
+        parse_meta_list_after_path(path, input).map(Meta::List)
+    } else if input.peek(Token![=]) {
+        parse_meta_name_value_after_path(path, input).map(Meta::NameValue)
+    } else {
+        Ok(Meta::Path(path))
+    }
+}
+
+fn parse_meta_list_after_path<'a>(path: Path, input: ParseStream<'a>) -> syn::Result<MetaList> {
+    let (delimiter, tokens) = input.step(|cursor| {
+        if let Some((TokenTree::Group(g), rest)) = cursor.token_tree() {
+            let span = g.delim_span();
+            let delimiter = match g.delimiter() {
+                Delimiter::Parenthesis => MacroDelimiter::Paren(Paren(span)),
+                Delimiter::Brace => MacroDelimiter::Brace(Brace(span)),
+                Delimiter::Bracket => MacroDelimiter::Bracket(Bracket(span)),
+                Delimiter::None => {
+                    return Err(cursor.error("expected delimiter"));
+                }
+            };
+            Ok(((delimiter, g.stream()), rest))
+        } else {
+            Err(cursor.error("expected delimiter"))
+        }
+    })?;
+    Ok(MetaList {
+        path,
+        delimiter,
+        tokens,
+    })
+}
+
+fn parse_meta_name_value_after_path<'a>(
+    path: Path,
+    input: ParseStream<'a>,
+) -> syn::Result<MetaNameValue> {
+    let eq_token: Token![=] = input.parse()?;
+    let ahead = input.fork();
+    let lit: Option<Lit> = ahead.parse()?;
+    let value = if let (Some(lit), true) = (lit, ahead.is_empty()) {
+        input.advance_to(&ahead);
+        Expr::Lit(ExprLit {
+            attrs: Vec::new(),
+            lit,
+        })
+    } else if input.peek(Token![#]) && input.peek2(token::Bracket) {
+        return Err(input.error("unexpected attribute inside of attribute"));
+    } else {
+        input.parse()?
+    };
+    Ok(MetaNameValue {
+        path,
+        eq_token,
+        value,
+    })
+}
+
+#[derive(Debug, Clone)]
+// Addressing this would break many users of the crate.
+#[allow(clippy::large_enum_variant)]
+pub enum NestedMeta {
+    Meta(syn::Meta),
+    Lit(syn::Lit),
+}
+
+impl NestedMeta {
+    pub fn parse_meta_list(tokens: TokenStream) -> syn::Result<Vec<Self>> {
+        syn::punctuated::Punctuated::<NestedMeta, Token![,]>::parse_terminated
+            .parse2(tokens)
+            .map(|punctuated| punctuated.into_iter().collect())
+    }
+}
+
+impl syn::parse::Parse for NestedMeta {
+    fn parse(input: syn::parse::ParseStream<'_>) -> syn::Result<Self> {
```

Comment:
This implementation needs some comments explaining why it exists and how it differs from the one in `syn` for the two variants.